distributed/client.py

```diff
@@ -2311,22 +2311,26 @@ async def _gather_remote(self, direct, local_worker):
             self._gather_keys = None  # clear state, these keys are being sent off
             self._gather_future = None
 
-            if direct or local_worker:  # gather directly from workers
-                who_has = await retry_operation(self.scheduler.who_has, keys=keys)
-                data2, missing_keys, missing_workers = await gather_from_workers(
-                    who_has, rpc=self.rpc, close=False
-                )
-                response = {"status": "OK", "data": data2}
-                if missing_keys:
-                    keys2 = [key for key in keys if key not in data2]
-                    response = await retry_operation(self.scheduler.gather, keys=keys2)
-                    if response["status"] == "OK":
-                        response["data"].update(data2)
+            if not direct and not local_worker:
+                # ask scheduler to gather data for us
+                return await retry_operation(self.scheduler.gather, keys=keys)
+
+            # gather directly from workers
+            async def who_has(keys: Iterable[str]) -> dict[str, Collection[str]]:
+                return await retry_operation(self.scheduler.who_has, keys=list(keys))
 
-            else:  # ask scheduler to gather data for us
-                response = await retry_operation(self.scheduler.gather, keys=keys)
+            data, missing_keys = await gather_from_workers(
+                keys=keys, who_has=who_has, rpc=self.rpc
+            )
+            response: dict[str, Any] = {"status": "OK", "data": data}
+            if missing_keys:
+                response = await retry_operation(
+                    self.scheduler.gather, keys=missing_keys
+                )
+                if response["status"] == "OK":
+                    response["data"].update(data)
```

Comment:
- #7993 
This PR does not change the previous behaviour.

Reply:
I cannot see any aesthetical refactoring in this PR that could be moved to a separate PR while preserving the functionality.